compiler/rustc_middle/src/ty/error.rs

```diff
@@ -497,13 +497,18 @@ impl<T> Trait<T> for X {
                             "consider constraining the associated type `{}` to `{}`",
                             values.found, values.expected,
                         );
-                        if !self.suggest_constraint(
+                        if !(self.suggest_constraining_opaque_associated_type(
+                            db,
+                            &msg,
+                            proj_ty,
+                            values.expected,
+                        ) || self.suggest_constraint(
```

Comment:
I added this to avoid losing a helpful [suggestion](https://github.com/rust-lang/rust/blob/b88e06c82084f81df5a730881d4f52f475d8abc7/src/test/ui/associated-types/impl-trait-return-missing-constraint.stderr#L12) in the impl-trait-return-missing-constraint test.

I wonder how much else of the `expected_projection` function makes sense to run in this case (when the projection type is *found*) as well.